File: src/main/java/com/github/simiacryptus/aicoder/actions/code/CustomEditAction.java
package com.github.simiacryptus.aicoder.actions.code;

import com.github.simiacryptus.aicoder.util.ComputerLanguage;
import com.github.simiacryptus.aicoder.config.AppSettingsState;
import com.github.simiacryptus.aicoder.openai.CompletionRequest;
import com.github.simiacryptus.aicoder.util.IndentedText;
import com.github.simiacryptus.aicoder.util.UITools;
import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.CommonDataKeys;
import com.intellij.openapi.editor.Caret;
import com.intellij.openapi.editor.CaretModel;
import com.intellij.openapi.editor.Editor;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;

import static java.util.Objects.requireNonNull;


/**
 * The CustomEditAction class is an IntelliJ action that allows users to edit computer language code.
 * When the action is triggered, a dialog box will appear prompting the user to enter an instruction.
 * The instruction will then be used to transform the selected code.
 *
 * To use the CustomEditAction, first select the code that you want to edit.
 * Then, select the action in the context menu.
 * A dialog box will appear, prompting you to enter an instruction.
 * Enter the instruction and press OK.
*/
public class CustomEditAction extends AnAction {

    @Override
    public void update(@NotNull AnActionEvent e) {
        e.getPresentation().setEnabledAndVisible(isEnabled(e));
        super.update(e);
    }

    private static boolean isEnabled(@NotNull AnActionEvent e) {
        if(!UITools.INSTANCE.hasSelection(e)) return false;
        if(null == ComputerLanguage.getComputerLanguage(e)) return false;
        return true;
    }

    @Override
    public void actionPerformed(@NotNull final AnActionEvent e) {
        final @NotNull Editor editor = e.getRequiredData(CommonDataKeys.EDITOR);
        final @NotNull CaretModel caretModel = editor.getCaretModel();
        final @NotNull Caret primaryCaret = caretModel.getPrimaryCaret();
        int selectionStart = primaryCaret.getSelectionStart();
        int selectionEnd = primaryCaret.getSelectionEnd();
        @Nullable String selectedText = primaryCaret.getSelectedText();
        @NotNull String computerLanguage = requireNonNull(ComputerLanguage.getComputerLanguage(e)).name();
        String instruction = JOptionPane.showInputDialog(null, "Instruction:", "Edit Code", JOptionPane.QUESTION_MESSAGE);
        AppSettingsState settings = AppSettingsState.getInstance();
        settings.addInstructionToHistory(instruction);
        @NotNull CompletionRequest request = settings.createTranslationRequest()
                .setInputType(computerLanguage)
                .setOutputType(computerLanguage)
                .setInstruction(instruction)
                .setInputAttribute("type", "before")
                .setOutputAttrute("type", "after")
                .setInputText(IndentedText.fromString(selectedText).getTextBlock())
                .buildCompletionRequest();
        @Nullable Caret caret = e.getData(CommonDataKeys.CARET);
        CharSequence indent = UITools.INSTANCE.getIndent(caret);
        UITools.INSTANCE.redoableRequest(request, indent, e,
                newText -> UITools.INSTANCE.replaceString(editor.getDocument(), selectionStart, selectionEnd, newText));
    }
}

File: src/main/kotlin/com/github/simiacryptus/aicoder/actions/code/CustomEditAction.kt
package com.github.simiacryptus.aicoder.actions.code

import com.github.simiacryptus.aicoder.actions.SelectionAction
import com.github.simiacryptus.aicoder.config.AppSettingsState
import com.github.simiacryptus.aicoder.util.UITools
import com.intellij.openapi.actionSystem.ActionUpdateThread
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.DataKey
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.project.Project
import com.simiacryptus.jopenai.models.chatModel
import com.simiacryptus.jopenai.proxy.ChatProxy
import java.awt.Component
import javax.swing.JOptionPane

/**
 * Action that allows custom editing of code selections using AI.
 * Supports multiple languages and provides custom edit instructions.
 */

open class CustomEditAction : SelectionAction<String>() {
    private val log = Logger.getInstance(CustomEditAction::class.java)
    override fun getActionUpdateThread() = ActionUpdateThread.BGT

    interface VirtualAPI {
        fun editCode(
            code: String,
            operation: String,
            computerLanguage: String,
            humanLanguage: String
        ): EditedText

        data class EditedText(
            var code: String? = null,
            var language: String? = null
        )
    }

    val proxy: VirtualAPI
        get() {
            val chatProxy = ChatProxy(
                clazz = VirtualAPI::class.java,
                api = api,
                temperature = AppSettingsState.instance.temperature,
                model = AppSettingsState.instance.smartModel.chatModel(),
            )
            chatProxy.addExample(
                VirtualAPI.EditedText(
                    """
                // Print Hello, World! to the console
                println("Hello, World!")
                """.trimIndent(),
                    "java"
                )
            ) {
                it.editCode(
                    """println("Hello, World!")""",
                    "Add code comments",
                    "java",
                    "English"
                )
            }
            return chatProxy.create()
        }

    override fun getConfig(project: Project?): String? {
        return UITools.showInputDialog(
            project as Component?,
            "Enter edit instruction:",
            "Edit Code",
            JOptionPane.QUESTION_MESSAGE
        ) as String?
    }

    override fun processSelection(state: SelectionState, instruction: String?): String {
        if (instruction == null || instruction.isBlank()) return state.selectedText ?: ""
        return try {
            UITools.run(state.project, "Processing Edit", true) { progress ->
                progress.isIndeterminate = true
                progress.text = "Applying edit: $instruction"
                val settings = AppSettingsState.instance
                val outputHumanLanguage = settings.humanLanguage
                settings.getRecentCommands("customEdits").addInstructionToHistory(instruction)
                val result = proxy.editCode(
                    state.selectedText ?: "",
                    instruction,
                    state.language?.name ?: "text",
                    outputHumanLanguage
                )
                result.code ?: state.selectedText ?: ""
            }
        } catch (e: Exception) {
            log.error("Failed to process edit", e)
            UITools.showErrorDialog(
                state.project,
                "Failed to process edit: ${e.message}",
                "Edit Error"
            )
            state.selectedText ?: ""
        }
    }

    override fun isEnabled(event: AnActionEvent): Boolean {
        if (!super.isEnabled(event)) return false
        return event.getData(DATA_KEY) != null
    }

    companion object {
        private val DATA_KEY = DataKey.create<String>("CustomEditAction.key")
    }
}